---
title: "Method Comparison: LME vs. EdgeR for Crossover Design"
subtitle: "BIOS658 Final Project"
author: "Amanda Robinson"
date: "December 1, 2025"
output:
  html_document:
    code_folding: show
    toc: true
    toc_float:
      collapsed: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, error = FALSE, 
                      message = FALSE)
```

# Introduction

# Methods

# Results

## Quality Assessment

```{r data}
library(tidyverse)
library(GEOquery)
library(pheatmap)
library(limma)
library(edgeR)

#To track figure and table numbers
fig.num <- 0
tab.num <- 0

#Data obtained from https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE231347
#read_csv() automatically unzips .gz files
gene_counts <- read_csv("GSE231347_counts.csv.gz")

gse <- getGEO("GSE231347", GSEMatrix = TRUE)
#Extract the expression set
gse_data <- gse[[1]]

# Create data frame from GEO phenoData.
phenotype <- pData(gse_data) %>%
  rename(condition = `condition:ch1`, 
         mdd_phenotype = `mdd phenotype:ch1`, 
         timepoint = `time point:ch1`) %>%
  mutate(id = str_extract(title, "G[0-9]+")) %>%
  filter(timepoint == "T105") %>%
  dplyr::select(title, id, mdd_phenotype, condition)

gene_counts <- gene_counts[, c("gene", phenotype$title)]
```

```{r qa-library-size}
library_size <- gene_counts %>%
  dplyr::select(-gene) %>%
  #Count the number of reads per sample
  colSums()

#Summary statistics for text
lib_summ <- round(quantile(library_size, c(0.25, 0.5, 0.75))/1000000, 1)
```

```{r qa-library-size-fig}
fig.num <- fig.num + 1

ggplot() + 
  geom_boxplot(aes(y = library_size/1000000)) +
  ylab("Library Size (Millions)") +
  ggtitle(paste0("Figure ", fig.num, ": Distribution of Library Size")) +
  theme_bw()
```

```{r qa-count-distribution}
gene_counts_long <- gene_counts %>%
  pivot_longer(-gene, names_to = "sample", values_to = "count") %>%
  mutate(log2_count = log2(count + 1))

log2_q3_range <- gene_counts_long %>%
  group_by(sample) %>%
  summarise(q3 = quantile(log2_count, 0.75)) %>%
  summarise(min = min(q3), 
            max = max(q3))
```

```{r qa-count-distribution-fig}
fig.num <- fig.num + 1

gene_counts_long %>%
  ggplot() + 
  geom_boxplot(aes(y = log2_count, x = sample)) +
  xlab("Sample") +
  ylab("Log2 Count + 1") +
  ggtitle(paste0("Figure ", fig.num, ": Distribution of Gene Count by Sample")) +
  theme_bw() +
  theme(axis.text.x = element_blank(), 
        axis.ticks.x = element_blank())
```

```{r housekeeping}
gene_counts <- gene_counts %>%
  dplyr::select(-gene) %>%
  as.matrix()

# phenotype$title == colnames(gene_counts)
```


```{r qa-sample-correlation}

```

## Preprocessing

```{r preprocessing-no-id}
TypeCondition <- interaction(phenotype$mdd_phenotype, phenotype$condition)

#No ID
#We will also need the design matrix
design <- model.matrix(~ TypeCondition)

#A. Format data as a DGEList
#We need to create a DGEList object to feed into the edgeR functions. 
y <- DGEList(counts = gene_counts, genes = rownames(gene_counts), 
             group = TypeCondition)

#B. Filter
#The documentation recommends filtering out genes with low or zero counts across
#all samples using this code. It also says you can include variables of 
#scientific interest (group in our case) in this step, which I choose to do 
#here.
keep <- filterByExpr(y, group = TypeCondition)
#We need to recalculate library size since we removed some genes
y <- y[keep, , keep.lib.sizes = FALSE]

#C. Normalization
#A few highly expressed genes may account for most of the library size, 
#resulting in under-sampling of genes with lower expression. We can address this
#issue through normalization. 
y <- normLibSizes(y)

#D. Estimating Dispersion
#The authors recommend calculating the overall and tag-wise dispersion
y <- estimateDisp(y_test, design = design)
```

## Batch Effect Correction

## Differential Expression Analysis

```{r differential-expression}
#E. Fit and Test Model
#I am using the quasi-likelihood pipeline
#"The QL F-tests provide more rigorous error rate control compared to methods 
#that do not fully account for the uncertainty in quasi-dispersion estimation."
fit_edgeR <- glmQLFit(y, design)
```

## Functional Enrichment Analysis

# Discussion

# Bibliography
